### PR TITLE
【Hackathon 4th No.29】为 Paddle 新增 paddle.sparse.slice 稀疏 API

### PR DESCRIPTION
**Pull request by ScottWong98**
### PR types

New features

### PR changes

APIs

### Description

实现第四期 Hackathon [第 29 个任务](https://github.com/PaddlePaddle/community/blob/master/hackthon_4th/%E3%80%90PaddlePaddle%20Hackathon%204%E3%80%91%20%E6%A0%B8%E5%BF%83%E6%A1%86%E6%9E%B6%E5%BC%80%E6%BA%90%E8%B4%A1%E7%8C%AE%20API%20%E5%BC%80%E5%8F%91%E4%BB%BB%E5%8A%A1%E5%90%88%E9%9B%86.md#task29)

中文文档: https://github.com/PaddlePaddle/docs/pull/5895 

原始 RFC: https://github.com/PaddlePaddle/community/pull/382

原始 RFC 的修复: https://github.com/PaddlePaddle/community/pull/539

### CODE REVIEW

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+                                    const int64_t* starts,
+                                    const int64_t* ends,
+                                    const int64_t axes_size,
+                                    const int64_t sparse_dim,
+                                    const int64_t x_nnz,
+                                    const int64_t out_nnz,
+                                    int64_t* out_indices_data,
+                                    T* out_values_data) {
+  int64_t tid = static_cast<int64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  if (tid == 0) {
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 这里只用一个线程， 如果nnz比较大的情况这里会是性能瓶颈，所以这里需要并行下。

Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 感谢您的 Review，已修改完毕，麻烦重新 Review 一下

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+
+  // copy d_out_nnz from device to host (out_nnz)
+  int32_t out_nnz = 0;
+  phi::backends::gpu::GpuMemcpyAsync(&out_nnz,
+                                     d_out_nnz_ptr,
+                                     sizeof(int32_t),
+                                     gpuMemcpyDeviceToHost,
+                                     dev_ctx.stream());
+  // sort `d_out_nnz_indices_ptr`
+  d_out_nnz_indices.Resize({out_nnz});
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu:178]: GpuMemcpyAsync是异步的，需要加个dev_ctx.Wait()保证out_nnz拿到数据了

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+  thrust::device_vector<int64_t> d_out_nnz_indices_vec(
+      d_out_nnz_indices_ptr, d_out_nnz_indices_ptr + out_nnz);
+  thrust::sort(d_out_nnz_indices_vec.begin(), d_out_nnz_indices_vec.end());
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 调用thrust需要把stream信息传进去

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+  phi::backends::gpu::GpuMemcpyAsync(
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 这里是先排序，然后再copy一次？应该可以直接对d_out_nnz_indices_ptr进行排序

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+      phi::Stream(reinterpret_cast<phi::StreamId>(dev_ctx.stream())));
+  int64_t* d_ends = reinterpret_cast<int64_t*>(d_ends_tensor->ptr());
+  memory_utils::Copy(dev_ctx.GetPlace(),
+                     d_ends,
+                     phi::CPUPlace(),
+                     ends.data(),
+                     sizeof(int64_t) * ends.size(),
+                     dev_ctx.stream());
+
+  const auto* x_indices_data = x.indices().data<int64_t>();
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 当前indices的类型需要支持int32和int64等多种类型，可以参考其他算子，对kernel做个二次分发

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+                                          const int64_t min_col,
+                                          const int64_t max_col,
+                                          int64_t* out_crows_data,
+                                          int64_t* out_cols_data,
+                                          T* out_values_data,
+                                          const int64_t x_cols_offset = 0,
+                                          const int64_t out_crows_offset = 0,
+                                          const int64_t out_cols_offset = 0) {
+  int64_t tid = static_cast<int64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  if (tid == 0) {
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 这里也只有一个线程，也优化下吧

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+                                                      starts[1],
+                                                      ends[1],
+                                                      d_out_nnz_ptr,
+                                                      0);
+  int32_t out_nnz = 0;
+  phi::backends::gpu::GpuMemcpyAsync(&out_nnz,
+                                     d_out_nnz_ptr,
+                                     sizeof(int32_t),
+                                     gpuMemcpyDeviceToHost,
+                                     dev_ctx.stream());
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu:336]: 同样的，在使用out_nnz之前需要加个wait。

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+                      const std::vector<int64_t>& starts,
+                      const std::vector<int64_t>& ends,
+                      const phi::DDim& out_dims,
+                      SparseCsrTensor* out) {
+  const auto* x_crows_data = x.crows().data<int64_t>();
+  const auto* x_cols_data = x.cols().data<int64_t>();
+  const auto* x_values_data = x.values().data<T>();
+  const int64_t x_dim0 = x.dims()[0], x_n_rows = x.dims()[1];
+  // copy x_crows_data from device to host
+  int64_t* temp_x_crows_data = new int64_t[x_dim0 * (x_n_rows + 1)];
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 可以用vector或者tensor

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+                                     dev_ctx.stream());
+
+  // Step1: Get the number of non zero elements for out
+  DenseTensor d_nnz = phi::Empty<int32_t>(dev_ctx, {1});
+  int* d_nnz_ptr = d_nnz.data<int32_t>();
+
+  std::vector<int64_t> all_nnzs(ends[0] - starts[0]);
+  int64_t x_cols_offset = 0, out_nnz = 0;
+
+  for (int64_t i = 0; i < x_dim0; ++i) {
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 可否把这一维的处理也并行化？

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+  const int64_t out_dim0 = out_dims[0], out_n_rows = out_dims[1];
+  DenseTensor out_crows =
+      phi::Empty<int64_t, Context>(dev_ctx, {out_dim0 * (out_n_rows + 1)});
+  DenseTensor out_cols = phi::Empty<int64_t, Context>(dev_ctx, {out_nnz});
+  DenseTensor out_values = phi::Empty<T, Context>(dev_ctx, {out_nnz});
+  out->SetMember(out_crows, out_cols, out_values, out_dims);
+
+  x_cols_offset = 0;
+  int64_t out_crows_offset = 0, out_cols_offset = 0;
+  for (int64_t i = 0; i < x_dim0; ++i) {
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 同上

File: python/paddle/fluid/tests/unittests/test_sparse_slice_op.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+# limitations under the License.
+
+import random
+import unittest
+
+import numpy as np
+
+import paddle
+
+# To ensure that the result after slicing is not a sparse tensor with all zeros.
```
Comment (zkh2016) [inline python/paddle/fluid/tests/unittests/test_sparse_slice_op.py]: 当前应该已经允许全零的情况，可以修改下单测

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 2 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu:178]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 3 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 4 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 5 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 6 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 7 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu:336]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 9 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 10 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: Done

File: python/paddle/fluid/tests/unittests/test_sparse_slice_op.py (hunk 11 above)
Comment (ScottWong98) [reply python/paddle/fluid/tests/unittests/test_sparse_slice_op.py]: Done

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu (hunk 8 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu]: 目前修改了csr的slice逻辑，`temp_x_crows_data` 已删除

**Comment by ScottWong98** [comment]
@zkh2016 您好，目前已经修改完毕，麻烦再重新Review以下

File: paddle/phi/kernels/sparse/gpu/slice_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,620 @@
...
+  int32_t out_nnz = 0;
+  phi::backends::gpu::GpuMemcpyAsync(&out_nnz,
+                                     d_out_nnz_ptr,
+                                     sizeof(int32_t),
+                                     gpuMemcpyDeviceToHost,
+                                     dev_ctx.stream());
+  dev_ctx.Wait();
+  // sort `d_out_nnz_indices_ptr`
+  d_out_nnz_indices.Resize({out_nnz});
+  thrust::sort(thrust::cuda::par.on(dev_ctx.stream()),
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/slice_kernel.cu:182]: ROCM的CI挂了，这里可以判断下，
```
#ifdef PADDLE_WITH_HIP
thrust::hip::par.on(dev_ctx.stream())
#else  
thrust::cuda::par.on(dev_ctx.stream())
#end 
```

Comment (ScottWong98) [reply paddle/phi/kernels/sparse/gpu/slice_kernel.cu:182]: Done

**Review by zkh2016 — DISMISSED** [review]

File: paddle/phi/kernels/sparse/unary_kernel.h [hunk 13]
```diff
@@ -225,5 +225,21 @@ SparseCsrTensor ReshapeCsr(const Context& dev_ctx,
   return csr;
 }
 
+template <typename T, typename Context>
+void SliceCooKernel(const Context& dev_ctx,
+                    const SparseCooTensor& x,
+                    const phi::IntArray& axes_arr,
+                    const phi::IntArray& starts_arr,
+                    const phi::IntArray& ends_arr,
```
Comment (jeff41404) [inline paddle/phi/kernels/sparse/unary_kernel.h]: Why use parameter name of `axes_arr`, `starts_arr` and `ends_arr`? No other kernel are not named like this, just using `axes`, `starts` and `ends` is better?

File: paddle/phi/kernels/sparse/unary_grad_kernel.h [hunk 14]
```diff
@@ -121,5 +121,22 @@ void ReshapeCsrGradKernel(const Context& dev_ctx,
                           const SparseCsrTensor& dout,
                           SparseCsrTensor* dx);
 
+template <typename T, typename Context>
+void SliceCooGradKernel(const Context& dev_ctx,
+                        const SparseCooTensor& x,
+                        const SparseCooTensor& out_grad,
+                        const phi::IntArray& axes_arr,
+                        const phi::IntArray& starts_arr,
+                        const phi::IntArray& ends_arr,
```
Comment (jeff41404) [inline paddle/phi/kernels/sparse/unary_grad_kernel.h]: Why use parameter name of `axes_arr`, `starts_arr` and `ends_arr`? No other kernel are not named like this, just using `axes`, `starts` and `ends` is better?

File: paddle/phi/kernels/sparse/unary_grad_kernel.h [hunk 15]
```diff
@@ -121,5 +121,22 @@ void ReshapeCsrGradKernel(const Context& dev_ctx,
...
+                        SparseCooTensor* x_grad);
+
+template <typename T, typename Context>
+void SliceCsrGradKernel(const Context& dev_ctx,
+                        const SparseCsrTensor& x,
+                        const SparseCsrTensor& out_grad,
+                        const phi::IntArray& axes_arr,
+                        const phi::IntArray& starts_arr,
+                        const phi::IntArray& ends_arr,
```
Comment (jeff41404) [inline paddle/phi/kernels/sparse/unary_grad_kernel.h]: same issue as above

File: paddle/phi/kernels/sparse/unary_kernel.h [hunk 16]
```diff
@@ -225,5 +225,21 @@ SparseCsrTensor ReshapeCsr(const Context& dev_ctx,
...
+                    SparseCooTensor* out);
+
+template <typename T, typename Context>
+void SliceCsrKernel(const Context& dev_ctx,
+                    const SparseCsrTensor& x,
+                    const phi::IntArray& axes_arr,
+                    const phi::IntArray& starts_arr,
+                    const phi::IntArray& ends_arr,
```
Comment (jeff41404) [inline paddle/phi/kernels/sparse/unary_kernel.h]: same issue as above

File: paddle/phi/kernels/sparse/unary_kernel.h (hunk 13 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_kernel.h]: Thank you very much for your review! 

The reason is that the data type of these parameters is `IntArray`, and I will get their underlying data in implementing the function, like `axes_arr.GetData()`. For convenience, I use `axes_arr` as the parameter's name and `axes` as the variable name indicating the underlying data described above in the function body.

And I found that the names of the `starts` and `ends` parameters are also `starts_array` and `ends_array` in DenseTensor's slice function:

https://github.com/PaddlePaddle/Paddle/blob/d1d43c22c81e12b825cfbadb304435deda0f2389/paddle/phi/kernels/impl/slice_kernel_impl.h#L102-L114

File: paddle/phi/kernels/sparse/unary_grad_kernel.h (hunk 14 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_grad_kernel.h]: same reason as above

File: paddle/phi/kernels/sparse/unary_grad_kernel.h (hunk 15 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_grad_kernel.h]: same reason as above

File: paddle/phi/kernels/sparse/unary_kernel.h (hunk 16 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_kernel.h]: same reason as above

File: paddle/phi/kernels/sparse/unary_kernel.h (hunk 13 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_kernel.h]: Additionally, the configuration of DenseTensor's slice API in `legacy_ops.yaml` is:

https://github.com/PaddlePaddle/Paddle/blob/cdbf62f829a245bc36ca6db93d2f3c838a1ce64a/paddle/phi/api/yaml/legacy_ops.yaml#L906-L913

**Comment by ScottWong98** [comment]
@jeff41404 Thank you again for your review. 

Finally, it might be better to use `axes`, `starts`, and `ends` as parameter names. And I have changed `axes_arr`, `starts_arr`, and `ends_arr` to `axes`, `starts`, and `ends`. 

Please see the latest commit for details and review it again!

File: python/paddle/sparse/unary.py [hunk 17]
```diff
@@ -836,3 +836,87 @@ def isnan(x, name=None):
...
+    the axis :math:`i-1` (here 0 is the initial position).
+    If the value passed to ``starts`` or ``ends`` is greater than the number of elements
+    in the dimenstion (n), it represents n.
+    For slicing to the end of a dimension with unknown size, it is recommended to pass
+    in INT_MAX. The size of ``axes`` must be equal to ``starts`` and ``ends``.
+
+    Args:
+        x (Tensor): The input Tensor (``SparseCooTensor`` or ``SparseCsrTensor``), it's data type should be ``float16``, ``float32``, ``float64``, ``int32``, ``int64``.
+        axes (list|tuple|Tensor): The data type is ``int32``.If ``axes`` is a list or tuple, the elements of
+                it should be integers or Tensors with shape [1]. If ``axes`` is a Tensor, it should be a 1-D Tensor.
```
Comment (sunzhongkai588) [inline python/paddle/sparse/unary.py:857]: 请确认是否是 `shape [1]`，中文文档写着的是`形状为[]`，两者不对应。下同

Comment (ScottWong98) [reply python/paddle/sparse/unary.py:857]: 感谢您的 Review！此处确定是 `shape[1]`，中文文档已经修改，与此处对应。

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs

File: paddle/phi/kernels/sparse/unary_kernel.h (hunk 13 above)
Comment (jeff41404) [reply paddle/phi/kernels/sparse/unary_kernel.h]: The signature of the SliceKernel should mainly refers to https://github.com/PaddlePaddle/Paddle/blob/d1d43c22c81e12b825cfbadb304435deda0f2389/paddle/phi/kernels/slice_kernel.h#L25-L32

The function mentioned above in Paddle/paddle/phi/kernels/impl/slice_kernel_impl.h is its implementation, and it's better to be the same, If you are interested, can propose another PR to correct the parameter names in slice_kernel_impl.h

**Review by jeff41404 — APPROVED** [review]
LGTM

File: paddle/phi/kernels/sparse/unary_kernel.h (hunk 13 above)
Comment (ScottWong98) [reply paddle/phi/kernels/sparse/unary_kernel.h]: Thanks for your reply! I will propose a PR to correct it when I have free time.

**Pull request merged**